runtime/parachains/src/scheduler_common/mod.rs

```diff
@@ -0,0 +1,344 @@
+//! - Parachains that have a candidate pending availability in this fork of the chain
+//!   should not be assigned.
+//! - Validator assignments should not be gameable. Malicious cartels should not be able to
+//!   manipulate the scheduler to assign themselves as desired.
+//! - High or close to optimal throughput of parachains and parathreads. Work among validator groups should be balanced.
+//!
+//! The Scheduler manages resource allocation using the concept of "Availability Cores".
+//! There will be one availability core for each parachain, and a fixed number of cores
+//! used for multiplexing parathreads. Validators will be partitioned into groups, with the same
+//! number of groups as availability cores. Validator groups will be assigned to different availability cores
+//! over time.
+
+use frame_support::pallet_prelude::*;
+use primitives::{
+	CollatorId, CoreIndex, CoreOccupied, GroupIndex, Id as ParaId, ParathreadClaim,
+	ParathreadEntry, ScheduledCore,
+};
+use scale_info::TypeInfo;
+use sp_std::{collections::btree_map::BTreeMap, prelude::*};
+
+use crate::initializer::SessionChangeNotification;
+
+//#[cfg(test)]
+//mod tests;
+
+/// Reasons a core might be freed
+#[derive(Clone, Copy)]
+pub enum FreedReason {
+	/// The core's work concluded and the parablock assigned to it is considered available.
+	Concluded,
+	/// The core's work timed out.
+	TimedOut,
+}
+
+/// The assignment type.
+#[derive(Clone, Encode, Decode, TypeInfo)]
+#[cfg_attr(feature = "std", derive(PartialEq, Debug))]
+pub enum AssignmentKind {
+	/// A parachain.
+	Parachain,
+	/// A parathread.
+	Parathread(CollatorId, u32),
+}
+
+/// How a free core is scheduled to be assigned.
+#[derive(Clone, Encode, Decode, TypeInfo)]
+#[cfg_attr(feature = "std", derive(PartialEq, Debug))]
+pub struct CoreAssignment {
+	/// The core that is assigned.
+	pub core: CoreIndex,
+	/// The unique ID of the para that is assigned to the core.
+	pub para_id: ParaId,
+	/// The kind of the assignment.
+	pub kind: AssignmentKind,
+	/// The index of the validator group assigned to the core.
+	pub group_idx: GroupIndex,
+}
+
+impl CoreAssignment {
+	/// Get the ID of a collator who is required to collate this block.
+	pub fn required_collator(&self) -> Option<&CollatorId> {
+		match self.kind {
+			AssignmentKind::Parachain => None,
+			AssignmentKind::Parathread(ref id, _) => Some(id),
+		}
+	}
+
+	/// Get the `CoreOccupied` from this.
+	pub fn to_core_occupied(&self) -> CoreOccupied {
+		match self.kind {
+			AssignmentKind::Parachain => CoreOccupied::Parachain,
+			AssignmentKind::Parathread(ref collator, retries) =>
+				CoreOccupied::Parathread(ParathreadEntry {
+					claim: ParathreadClaim(self.para_id, collator.clone()),
+					retries,
+				}),
+		}
+	}
+}
+
+// TODO: need to link CoreOccupied variants to CoreAssigner impls
+// TOOD: Change interface to remove CoreIndex and GroupIndex to not leak abstraction
+pub trait CoreAssigner<T: crate::scheduler::pallet::Config> {
```

Comment:
As these are very likely to change, I don't think that would make much sense. The point of this PR is to make sure the way I decompose the scheduler is FRAMEy

Reply:
We do need docs before merge. In the meantime, it would be helpful to proactively leave comments somewhere (either in the code or in GitHub discussion) for particular design aspects you believe should be looked at more closely.